server-spi/src/main/java/org/keycloak/models/ClientProvider.java

```diff
@@ -146,4 +146,12 @@ default ClientModel addClient(RealmModel realm, String clientId) {
      * @param realm Realm.
      */
     void removeClients(RealmModel realm);
+
+    /**
+     * This function is called when a role is removed; this serves for removing references from clients to roles.
+     *
+     * @param realm Realm.
+     * @param role Role which will be removed.
+     */
+    void preRemoveClientProvider(RealmModel realm, RoleModel role);
```

Comment:
Can this be plain `preRemove` similarly to other `preRemove` methods?

Reply:
@hmlnarik If same name `preRemove`is used both legacy implementations `JpaRealmProvider` and `RealmCacheSession`, the same method should be used for all the implementations, so I needed to include the code for clients in the same one used for groups. I detected that the method was called several times because (x4, x2 becase it was called twice [here](https://github.com/keycloak/keycloak/pull/7815/files#diff-30940e65a4a38797c21073cfeed4d0e4beb58042ed778364f30468571d55a50dR335-R336), and another x2 when the same is done in `RealmCacheSession`). So I decided to rename the method to other signature.